File: backend/app/schemas/lead.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import LeadStatus, SignalPriority


class LeadCreate(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    summary: str | None = None
    source_url: str | None = Field(default=None, max_length=2048)
    source_label: str | None = Field(default=None, max_length=255)
    priority: SignalPriority | None = None
    lead_score: int | None = Field(default=None, ge=0, le=100)
    recommendation: str | None = None
    notes: str | None = None
    extracted_data: dict[str, Any] = Field(default_factory=dict)


class LeadUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=500)
    summary: str | None = None
    source_url: str | None = Field(default=None, max_length=2048)
    source_label: str | None = Field(default=None, max_length=255)
    priority: SignalPriority | None = None
    lead_score: int | None = Field(default=None, ge=0, le=100)
    recommendation: str | None = None
    notes: str | None = None
    status: LeadStatus | None = None
    closed_reason: str | None = Field(default=None, max_length=500)


class LeadResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    signal_id: UUID | None
    created_by_id: UUID | None
    scrape_job_id: UUID | None
    scrape_result_id: UUID | None
    title: str
    summary: str | None
    source_url: str | None
    source_label: str | None
    priority: SignalPriority | None
    lead_score: int | None
    recommendation: str | None
    notes: str | None
    extracted_data: dict[str, Any]
    confidence_score: float | None
    status: LeadStatus
    status_updated_by_id: UUID | None
    status_updated_at: datetime | None
    closed_reason: str | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class PaginatedLeadsResponse(BaseModel):
    items: list[LeadResponse]
    total: int
    limit: int
    offset: int

File: frontend/src/types/lead.ts
export type LeadStatus = 'new' | 'contacted' | 'won' | 'lost' | 'archived'

export type SignalPriority = 'low' | 'medium' | 'high'

export interface Lead {
  id: string
  organization_id: string
  signal_id: string | null
  created_by_id: string | null
  scrape_job_id: string | null
  scrape_result_id: string | null
  title: string
  summary: string | null
  source_url: string | null
  source_label: string | null
  priority: SignalPriority | null
  lead_score: number | null
  recommendation: string | null
  notes: string | null
  extracted_data: Record<string, unknown>
  confidence_score: number | null
  status: LeadStatus
  status_updated_by_id: string | null
  status_updated_at: string | null
  closed_reason: string | null
  closed_at: string | null
  created_at: string
  updated_at: string
}

export interface PaginatedLeadsResponse {
  items: Lead[]
  total: number
  limit: number
  offset: number
}

export interface LeadListParams {
  status?: LeadStatus
  priority?: SignalPriority
  scrape_job_id?: string
  signal_id?: string
  limit?: number
  offset?: number
}

export interface LeadCreateRequest {
  title: string
  summary?: string | null
  source_url?: string | null
  source_label?: string | null
  priority?: SignalPriority | null
  lead_score?: number | null
  recommendation?: string | null
  notes?: string | null
  extracted_data?: Record<string, unknown>
}

export interface LeadUpdateRequest {
  title?: string
  summary?: string | null
  source_url?: string | null
  source_label?: string | null
  priority?: SignalPriority | null
  lead_score?: number | null
  recommendation?: string | null
  notes?: string | null
  status?: LeadStatus
  closed_reason?: string | null
}

export type LeadSortField =
  | 'title'
  | 'lead_score'
  | 'status'
  | 'source_label'
  | 'priority'
  | 'created_at'

export type SortDirection = 'asc' | 'desc'
